Fix part of #7450: modified email_manager.BulkEmailsTests to use public methods instead of private methods

brianlinUM opened the pull request:
## Overview
<!--
READ ME FIRST: Please answer *both* questions below and check off every point from the essential checklist!
-->

1. This PR fixes part of #[7450].
2. This PR does the following:

For the following tests in email_manager.BulkEmailsTests:

- 	test_that_correct_email_is_sent
- 	test_that_exception_is_raised_for_unauthorised_sender

modified them to use public methods instead of private methods.

To facilitate this, modified email_manager.send_user_query_email to take in testing
args.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

Commit:
Changed BulkEmailsTests to use public methods #7450

Tests changed in email_manager_test.BulkEmailsTests:
	test_that_correct_email_is_sent
	test_that_exception_is_raised_for_unauthorised_sender

Modified email_manager.send_user_query_email to take in testing
args.

brianlinUM commented:
Hello! Is this the sort of pattern you are looking for?

kevintab95 commented on core/domain/email_manager_test.py:
```diff
@@ -2422,12 +2422,13 @@ def test_that_correct_email_is_sent(self):
-                feconf.BULK_EMAIL_INTENT_MARKETING, email_subject,
-                email_html_body, self.SENDER_EMAIL, self.SENDER_USERNAME)
-            # pylint: enable=protected-access
+            email_manager.send_user_query_email(
```
Hi @brianlinUM, can you try to use the `send_user_query_email` method without adding extra parameters? We should avoid passing parameters like "test_only" to the actual method.

brianlinUM replied:
Hi @kevintab95, thank you for your feedback! I made some changes so that the test calls send_user_query_email directly ( I found out it has the same functionality as the private method). I also reverted send_user_query_email to its original state. I forgot to fetch/merge from upstream/develop before making changes, so commit 509e936 is the one in which I made changes. Would this be a problem?

Commit:
Modified BulkEmailsTests to use public methods directly

reverted last commit's changes and made the tests simply call
the public method (send_user_query_email) instead of the private
method since they have the same functionality.

Commit:
Fixed linting error

fixed email.manager to email_manager

Commit:
Merge remote-tracking branch 'upstream/develop' into fix-part-of-7450-email-manager-BulkEmailsTests

kevinlee12 commented:
@brianlinUM the backend test failure is due to lack of coverage, please look at which lines you are missing.

kevinlee12 commented:
Look at the bottom of the backend test logs to see the coverage report.

brianlinUM commented:
The issue is that in` _send_bulk_mail_in_transaction` of `email_manager._send_bulk_mail`:
```
if instance is None:
    instance_id = email_models.BulkEmailModel.get_new_id('')
```
is not being run. I think this can be fixed in send_user_query_email by removing:
```
bulk_email_model_id = email_models.BulkEmailModel.get_new_id('')
``` 
and have `_send_bulk_mail` handle the instance_id creation and return it to `send_user_query_email`. Currently, both `_send_bulk_mail` and `send_user_query_email` handle instance_id creation, which is redundant. How does this solution sound?

kevintab95 commented:
@aks681 PTAL! 

@brianlinUM, I think you may be right about instance_id being set twice. I would defer to @aks681 to advice since he is the code-owner. Thanks!

aks681 commented:
@brianlinUM Yeah, that's true. In that case, make sure that `_send_bulk_mail` is only called in `send_user_query_email` and not anywhere else currently. If so, you can go ahead and change the docstring of `_send_bulk_mail` to have the instance_id field be only str, and remove its default value.

Then, you can remove that line. Since the aim of this PR is to prevent external use of `_send_bulk_email` anyway, it should be fine as any internal function that calls it should then call it with instance_id field filled.

Other than this, I have no comments!

Commit:
Removed redundant get_new_id for bulk email model

Both _send_bulk_mail and send_user_query_email in email_manager
had lines for getting a new bulk mail instanceID. Modified so
that caller of _send_bulk_mail is responsible for passing in
a instanceID.

Should also fix the less than 100 percent coverage issue.

kevintab95 commented:
@aks681 PTAL!

aks681 commented on core/domain/email_manager.py:
```diff
@@ -383,16 +383,14 @@ def _send_bulk_mail(
     cleaned_plaintext_body = html_cleaner.strip_html_tags(raw_plaintext_body)
 
-    def _send_bulk_mail_in_transaction(instance_id=None):
+    def _send_bulk_mail_in_transaction(instance_id):
```
Edit the docstring for this function listing its arguments.

brianlinUM replied:
Done!

aks681 reviewed (commented):
Only one comment, else lgtm!

Commit:
Added docstring for method arguments

Commit:
Merge remote-tracking branch 'upstream/develop' into fix-part-of-7450-email-manager-BulkEmailsTests

kevinlee12 commented:
@brianlinUM if you are ready for the reviewer to take another pass, please @-mention them.

@aks681 ptal!

aks681 reviewed (approved):
Lgtm!

kevintab95 reviewed (approved):
LGTM, thanks @brianlinUM!
